oiosaml/src/main/java/dk/gov/oio/saml/session/SessionCleanerService.java

```diff
@@ -0,0 +1,64 @@
+package dk.gov.oio.saml.session;
+
+import dk.gov.oio.saml.config.Configuration;
+import dk.gov.oio.saml.service.OIOSAML3Service;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.servlet.http.HttpSession;
+import java.util.Timer;
+import java.util.TimerTask;
+
+public class SessionCleanerService {
+    private static final Logger log = LoggerFactory.getLogger(SessionCleanerService.class);
+
+    private boolean initialized = false;
+    private Timer cleanupTimer;
+
+    public SessionCleanerService(Configuration configuration) {
+    }
+
+    public synchronized void updateCleaner(HttpSession session) {
+        try {
+            if (!initialized) {
+                long maxInactiveIntervalSeconds = session.getMaxInactiveInterval() > 0 ?
+                        (long)session.getMaxInactiveInterval() :
+                        30L * 60L /* defaults to 30 minutes */;
+
+                initializeCleaner(maxInactiveIntervalSeconds);
+
+                initialized = true;
+            }
+        } catch (Exception e) {
+            log.error("Unable to start session cleaner", e);
+        }
+    }
+
+    private void initializeCleaner(final long maxInactiveIntervalSeconds) {
+        log.info("Starting session cleaner with timeout '{}'", maxInactiveIntervalSeconds);
+        if (cleanupTimer != null) {
+            stopCleaner();
+        }
+
+        cleanupTimer = new Timer("Session Cleanup");
```

Comment:
Consider replacing with ScheduledThreadPoolExecutor, a more secure (stability wise) and flexible solution that is not using threads from the application thread pool.

Reply:
Agreed - boy scout rule when we migrate from the old implementation.

Reply:
Refactored to ScheduledExecutorService and fix synchronization issue